lib/liquid/context.rb

```diff
@@ -164,25 +193,13 @@ def find_variable(key, raise_on_not_found: true)
       # This was changed from find() to find_index() because this is a very hot
       # path and find_index() is optimized in MRI to reduce object allocation
       index = @scopes.find_index { |s| s.key?(key) }
-      scope = @scopes[index] if index
 
-      variable = nil
-
-      if scope.nil?
-        @environments.each do |e|
-          variable = lookup_and_evaluate(e, key, raise_on_not_found: raise_on_not_found)
-          # When lookup returned a value OR there is no value but the lookup also did not raise
-          # then it is the value we are looking for.
-          if !variable.nil? || @strict_variables && raise_on_not_found
-            scope = e
-            break
-          end
-        end
+      variable = if index
+        lookup_and_evaluate(@scopes[index], key, raise_on_not_found: raise_on_not_found)
+      else
+        try_find_in_environments(key, raise_on_not_found: raise_on_not_found)
       end
 
-      scope ||= @environments.last || @scopes.last
```

Comment:
I'm struggling to understand what this line (which is not in the new implementation) was doing, since at that point we already searched in both `@environments` and `@scopes` and so I'd expect that whatever scope this assigns, the variable is not in there.

Reply:
Likewise. It seems like it should always return `nil`.

My best guess is that this is copy / pasted from variable _setting_ code, where we would need to know the scope in order to set the value in the right place.